ledger/narwhal/subdag/src/lib.rs

```diff
@@ -50,39 +108,64 @@ fn is_sequential<T>(map: &BTreeMap<u64, T>) -> bool {
-        // Iterate over the previous certificate IDs.
-        for previous_certificate_id in certificate.previous_certificate_ids() {
-            let Some(previous_certificate) = subdag
-                .get(&(certificate.round() - 1))
-                .and_then(|map| map.iter().find(|certificate| certificate.id() == *previous_certificate_id))
-            else {
-                // It is either ordered or below the GC round.
-                continue;
-            };
-            // Insert the previous certificate into the set of already ordered certificates.
-            if !already_ordered.insert(previous_certificate.id()) {
-                // If the previous certificate is already ordered, continue.
-                continue;
+macro_rules! sanity_check_subdag_with_dfs {
+    ($subdag:expr) => {{
+        use std::collections::HashSet;
+
+        // Initialize a map for the certificates to commit.
+        let mut commit = BTreeMap::<u64, IndexSet<_>>::new();
+        // Initialize a set for the already ordered certificates.
+        let mut already_ordered = HashSet::new();
+        // Initialize a buffer for the certificates to order, starting with the leader certificate.
+        let mut buffer = $subdag.iter().next_back().map_or(Default::default(), |(_, leader)| leader.clone());
+        // Iterate over the certificates to order.
+        while let Some(certificate) = buffer.pop() {
+            // Insert the certificate into the map.
+            commit.entry(certificate.round()).or_default().insert(certificate.clone());
+            // Iterate over the previous certificate IDs.
+            for previous_certificate_id in certificate.previous_certificate_ids() {
+                let Some(previous_certificate) = $subdag
+                    .get(&(certificate.round() - 1))
+                    .and_then(|map| map.iter().find(|certificate| certificate.id() == *previous_certificate_id))
+                else {
+                    // It is either ordered or below the GC round.
+                    continue;
+                };
+                // Insert the previous certificate into the set of already ordered certificates.
+                if !already_ordered.insert(previous_certificate.id()) {
+                    // If the previous certificate is already ordered, continue.
+                    continue;
+                }
+                // Insert the previous certificate into the buffer.
+                buffer.insert(previous_certificate.clone());
             }
-            // Insert the previous certificate into the buffer.
-            buffer.insert(previous_certificate.clone());
         }
-    }
-    // Return `true` if the subdag matches the commit.
-    &commit == subdag
+        // Return `true` if the subdag matches the commit.
+        &commit == $subdag
+    }};
+}
+
+/// Performs sanity checks on the subdag.
+macro_rules! sanity_check_subdag {
```

Comment:
Does this have to be a macro, or could we also write something like the following?

```suggestion
fn sanity_check_subdag<T>(subdag BTreeMap<u64, IndexSet<T>>) -> Result<()> {
```

I always try to avoid macros as they harder to read, at least for me.

Reply:
not my go-to either, but at least this one is rather simple; I tried to change it into a function again now, but wasn't able to